src/Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -84,8 +84,15 @@ protected MessageBody(HttpProtocol context)
                     {
                         foreach (var memory in readableBuffer)
                         {
+                            // REVIEW: This *could* be slower if 2 things are true
```

Comment:
This might be a breaking change that I don't mind reverting. We're calling a different version of WriteAsync here. On .NET Core 2.1 it will delegate to the correct WriteAsync call but I haven't through through all of the cases. 

@stephentoub Do you have any thoughts about the fact that CopyToAsync has to basically decide what overload of WriteAsync to call on the other Stream?

Reply:
Could this `foreach` be extension for `ReadOnlySequence<byte>`?
```csharp
WriteAsync(this Stream, ReadOnlySequence<byte>, CancellationToken)
WriteAsync(this PipeWriter, ReadOnlySequence<byte>, CancellationToken)
```

Reply:
> Do you have any thoughts about the fact that CopyToAsync has to basically decide what overload of WriteAsync to call on the other Stream?

For the most part we've decided it's ok to start calling the `{ReadOnly}Memory<byte>` overloads and have switched to doing so in a variety of places where it provides additional value, e.g. where we can better respect pinning, where we can avoid a task allocation for a synchronously completing ReadAsync, etc.  In general all of the reading methods on Stream are meant to provide the same core semantics, and similarly for the writing methods, so in theory there shouldn't be a real behavioral difference when we switch from calling one to calling the other.  And as you mention the base virtual implementation on Stream does the right thing.

There is one case where we've taken extra precautions to support this.  Consider a library with a type that derives from a Stream-derived type, like MemoryStream, for example a LimitMemoryStream that sets a bound on how much can be written to the Stream.  It does this by overriding all of the virtual methods on the type.  Now we introduce a new virtual method on Stream and override that on MemoryStream, but this derived LimitMemoryStream doesn't yet override it (and maybe can't, if it's defined in a netstandard20 implementation).  When we then switch to calling this new method on MemoryStream, we potentially bypass the extra functionality that was provided by the derived type.  As such, in cases like this where there's a notable concern, we've added a type check to validate that `this` is in fact a concrete MemoryStream rather than a derived type, and if it is derived, we instead delegate to the base implementation that'll defer to the existing virtual method, e.g.
https://github.com/dotnet/coreclr/blob/9ff75d7600aa59dcec3edc16332794cc7007590b/src/mscorlib/shared/System/IO/MemoryStream.cs#L686-L693
This has obvious downsides that such a derived type doesn't benefit from any perf benefits that overload provides, but it limits the compatibility concern.

We've also gone through and overridden the new `Memory<byte>`-based overloads on all of the relevant streams in coreclr/corefx.  Of course if we've missed any of note, please open an issue so it can be corrected asap :)